### PR TITLE
Fix for pre-bound snapshot empty source error

### PR DESCRIPTION
**Pull request by hakanmemisoglu**
* Handle pre-bound snapshots for update.

### CODE REVIEW

**Comment by hakanmemisoglu** [comment]
Fix for #94 
/assign @jingxu97 @xing-yang

**Comment by hakanmemisoglu** [comment]
I signed it

**Comment by xing-yang** [comment]
Thanks @hakanmemisoglu for the fix!  We'll take a look.

**Comment by xing-yang** [comment]
Can you add a unit test for this change?

**Comment by hakanmemisoglu** [comment]
@xing-yang Sure. Let me update the PR.

**Comment by hakanmemisoglu** [comment]
@xing-yang @jingxu97 
I have added the unit test for prebound snapshot. 
However, I also needed to change the function `checkandBindSnapshotContent` to return updated content.

File: pkg/controller/snapshot_controller.go [hunk 1]
```diff
@@ -560,16 +559,27 @@ func (ctrl *csiSnapshotController) checkandUpdateBoundSnapshotStatusOperation(sn
...
-	driverName, snapshotID, timestamp, size, readyToUse, err := ctrl.handler.CreateSnapshot(snapshot, volume, class.Parameters, snapshotterCredentials)
-	if err != nil {
-		glog.Errorf("checkandUpdateBoundSnapshotStatusOperation: failed to call create snapshot to check whether the snapshot is ready to use %q", err)
-		return nil, err
+
+	if snapshot.Spec.Source == nil {
+		glog.V(5).Infof("checkandUpdateBoundSnapshotStatusOperation: snapshot [%s] is pre-bound to content [%s]", snapshot.Name, content.Name)
+		readyToUse, timestamp, size, err = ctrl.handler.GetSnapshotStatus(content)
+		if err != nil {
+			return nil, err
```
Comment (xing-yang) [inline pkg/controller/snapshot_controller.go]: Log an error msg here.

File: pkg/controller/snapshot_controller.go [hunk 2]
```diff
@@ -560,16 +559,27 @@ func (ctrl *csiSnapshotController) checkandUpdateBoundSnapshotStatusOperation(sn
...
+		var driverName string
+		var snapshotID string
+		driverName, snapshotID, timestamp, size, readyToUse, err = ctrl.handler.CreateSnapshot(snapshot, volume, class.Parameters, snapshotterCredentials)
+		if err != nil {
+			glog.Errorf("checkandUpdateBoundSnapshotStatusOperation: failed to call create snapshot to check whether the snapshot is ready to use %q", err)
+			return nil, err
+		}
+		glog.V(5).Infof("checkandUpdateBoundSnapshotStatusOperation: driver %s, snapshotId %s, timestamp %d, size %d, readyToUse %t", driverName, snapshotID, timestamp, size, readyToUse)
 	}
-	glog.V(5).Infof("checkandUpdateBoundSnapshotStatusOperation: driver %s, snapshotId %s, timestamp %d, size %d, readyToUse %t", driverName, snapshotID, timestamp, size, readyToUse)
```
Comment (xing-yang) [inline pkg/controller/snapshot_controller.go]: You can still keep this debug msg at the end as you'll get those info for a pre-bound snapshot too.

File: pkg/controller/snapshot_controller.go [hunk 3]
```diff
@@ -560,16 +559,27 @@ func (ctrl *csiSnapshotController) checkandUpdateBoundSnapshotStatusOperation(sn
...
-	if err != nil {
-		return nil, fmt.Errorf("failed to get input parameters to create snapshot %s: %q", snapshot.Name, err)
-	}
-	driverName, snapshotID, timestamp, size, readyToUse, err := ctrl.handler.CreateSnapshot(snapshot, volume, class.Parameters, snapshotterCredentials)
-	if err != nil {
-		glog.Errorf("checkandUpdateBoundSnapshotStatusOperation: failed to call create snapshot to check whether the snapshot is ready to use %q", err)
-		return nil, err
+
+	if snapshot.Spec.Source == nil {
+		glog.V(5).Infof("checkandUpdateBoundSnapshotStatusOperation: snapshot [%s] is pre-bound to content [%s]", snapshot.Name, content.Name)
```
Comment (xing-yang) [inline pkg/controller/snapshot_controller.go]: At this point we don't know if this will succeed yet, so the msg should reflect that.

**Comment by xing-yang** [comment]
I tested this PR and it worked for me.
/lgtm

**Comment by xing-yang** [comment]
@jingxu97 PTAL

**Comment by xing-yang** [comment]
@hakanmemisoglu Please resolve merge conflicts.  I'll merge this after your rebase.  Thanks.

**Comment by hakanmemisoglu** [comment]
@xing-yang @jingxu97 
Just resolved the conflicts.

**Pull request merged**